data/account/participationRegistry.go

```diff
@@ -282,15 +309,55 @@ func (db *participationDB) initializeCache() error {
 	return nil
 }
 
-func (db *participationDB) Insert(record Participation) (id ParticipationID, err error) {
-	db.mutex.Lock()
-	defer db.mutex.Unlock()
-
-	id = record.ID()
-	if _, ok := db.cache[id]; ok {
-		return id, ErrAlreadyInserted
+func (db *participationDB) writeThread() {
+	needsFlush := false
+	var err error
+	var lastErr error
+	for {
+		var wr partDBWriteRecord
+		var chanOk bool
+		if needsFlush {
```

Comment:
I think this deserves a comment. It looks like the code can be simplified but there's a lot going on and all of it is important.
```
// The flush command locks the mutex, then sends a flush event to unblock the queue read.
// When everything is flushed, the default is blocked until flush calls flushDone.Wait.
// Finally flushDone.Broadcast notifies all calls to Flush that the queue is now empty.
```